x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/rank/textsimilarity/TextSimilarityRankFeaturePhaseRankCoordinatorContext.java

```diff
@@ -0,0 +1,99 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+package org.elasticsearch.xpack.inference.rank.textsimilarity;
+
+import org.elasticsearch.action.ActionListener;
+import org.elasticsearch.client.internal.Client;
+import org.elasticsearch.inference.InferenceServiceResults;
+import org.elasticsearch.inference.InputType;
+import org.elasticsearch.inference.TaskType;
+import org.elasticsearch.search.rank.context.RankFeaturePhaseRankCoordinatorContext;
+import org.elasticsearch.search.rank.feature.RankFeatureDoc;
+import org.elasticsearch.xpack.core.inference.action.InferenceAction;
+import org.elasticsearch.xpack.core.inference.results.RankedDocsResults;
+
+import java.util.Arrays;
+import java.util.List;
+import java.util.Map;
+
+/**
+ * A {@code RankFeaturePhaseRankCoordinatorContext} that performs a rerank inference call to determine relevance scores for documents within
+ * the provided rank window.
+ */
+public class TextSimilarityRankFeaturePhaseRankCoordinatorContext extends RankFeaturePhaseRankCoordinatorContext {
+
+    protected final Client client;
+    protected final String inferenceId;
+    protected final String inferenceText;
+    protected final Float minScore;
+
+    public TextSimilarityRankFeaturePhaseRankCoordinatorContext(
+        int size,
+        int from,
+        int rankWindowSize,
+        Client client,
+        String inferenceId,
+        String inferenceText,
+        Float minScore
+    ) {
+        super(size, from, rankWindowSize);
+        this.client = client;
+        this.inferenceId = inferenceId;
+        this.inferenceText = inferenceText;
+        this.minScore = minScore;
+    }
+
+    @Override
+    protected void computeScores(RankFeatureDoc[] featureDocs, ActionListener<float[]> scoreListener) {
+        // Wrap the provided rankListener to an ActionListener that would handle the response from the inference service
+        // and then pass the results
+        final ActionListener<InferenceAction.Response> actionListener = scoreListener.delegateFailureAndWrap((l, r) -> {
+            float[] scores = extractScoresFromResponse(r);
+            assert scores.length == featureDocs.length;
```

Comment:
this should be a hard failure. Otherwise, we have no way of knowing which score corresponds to which doc.

Reply:
Updated logic to fail with exception.